tooling/nargo_fmt/src/visitor/item.rs

```diff
@@ -1,30 +1,91 @@
 use noirc_frontend::{
     parser::{Item, ItemKind},
     token::Token,
-    NoirFunction, ParsedModule,
+    Distinctness, NoirFunction, ParsedModule, Visibility,
 };
 
 use crate::utils::last_line_contains_single_line_comment;
 
 impl super::FmtVisitor<'_> {
     fn format_fn_before_block(&self, func: NoirFunction, start: u32) -> (String, bool) {
-        let slice = self.slice(start..func.span().start());
+        let name_span = func.name_ident().span();
+        let func_span = func.span();
 
-        let params_open = self
-            .span_before(func.name_ident().span().end()..func.span().start(), Token::LeftParen)
-            .start();
+        let mut result = self.slice(start..name_span.end()).to_owned();
+
+        let params_open =
+            self.span_before(name_span.end()..func_span.start(), Token::LeftParen).start();
 
         let last_span = if func.parameters().is_empty() {
-            params_open..func.span().start()
+            params_open..func_span.start()
         } else {
-            func.parameters().last().unwrap().1.span.unwrap().end()..func.span().start()
+            func.parameters().last().unwrap().span.end()..func_span.start()
         };
 
         let params_end = self.span_after(last_span, Token::RightParen).start();
 
-        let maybe_comment = self.slice(params_end..func.span().start());
+        let params_span = params_open..params_end;
+        let return_type_span = func.return_type().span;
+        let parameters = func.def.parameters;
+
+        if !func.def.generics.is_empty() {
+            let full_span = name_span.end()..params_open;
+            let start = name_span.end();
+            let end = self.span_after(full_span, Token::Greater).start();
+
+            let generics = crate::visitor::expr::format_expr_seq(
+                "<",
+                ">",
+                self.fork(),
+                false,
+                func.def.generics,
+                (start..end).into(),
+                super::expr::Tactic::LimitedHorizontalVertical(self.config.max_width),
+            );
```

Comment:
Can you try to fix the formatting for the two calls to `format_expr_seq` here? E.g. importing the function and the tactic, and creating intermediate `let`s such that the call can fit on one or two lines instead of 9?

Reply:
Couldn't reduce it to 2 lines, I think this part of the code needs to be refactored in another PR, perhaps using the builder pattern or something similar.